File: C++/5-BinarySearch/Problems/3-FindMinimum-in-RotatedSortedArray/FindMinimum-in-RotatedSortedArrayPro.js
/**
 * Binary Search: Find Minimum in Rotated Sorted Array
 *
 * Problem:
 *
 * Given a sorted array that has been rotated n number of times, find the minimum element in it.
 *
 * Example:
 *
 * Input: [3, 4, 5, 1, 2]
 * Output: 1
 *
 * Constraints:
 *
 * -   The array will not have duplicate elements.
 * -   The array is sorted but rotated.
 * -   The rotation is in the clockwise direction.
 *
 *
 * Approaches:
 *
 * 1.  **Linear Search:**
 * -   Iterate through the array and find the minimum element.
 * -   Time Complexity: O(n)
 * -   Space Complexity: O(1)
 *
 * 2.  **Binary Search (Optimized):**
 * -   Use binary search to find the minimum element.
 * -   The key idea is that the rotated array has two sorted parts.
 * -   We can use binary search to find the point where the two sorted parts meet, which is the minimum element.
 * -   Time Complexity: O(log n)
 * -   Space Complexity: O(1)
 *
 * 3.  **Recursive Binary Search:**
 * -   Implement the binary search approach using recursion.
 * -   Time Complexity: O(log n)
 * -   Space Complexity: O(log n) (due to the call stack)
 *
 * 4. **Binary Search with Duplicates Handling:**
 * - Handles cases where the rotated array contains duplicate elements.
 * - Time Complexity: O(log n) in the average case, O(n) in the worst case.
 * - Space Complexity: O(1)
 *
 * 5. **ES6 Optimized Binary Search**
 * - Uses modern ES6 syntax for a more concise implementation.
 * - Time Complexity: O(log n)
 * - Space Complexity: O(1)
 */

/**
 * 1. Linear Search Approach
 *
 * @param {number[]} nums The rotated sorted array.
 * @returns {number} The minimum element in the array.
 */
function findMinimumLinearSearch(nums) {
    if (!nums || nums.length === 0) {
        return undefined; // Handle empty array case
    }

    let min = nums[0];
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] < min) {
            min = nums[i];
        }
    }
    return min;
}

/**
 * 2. Binary Search (Optimized) Approach
 *
 * @param {number[]} nums The rotated sorted array.
 * @returns {number} The minimum element in the array.
 */
function findMinimumBinarySearch(nums) {
    if (!nums || nums.length === 0) {
        return undefined; // Handle empty array case
    }

    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        const mid = Math.floor((left + right) / 2);

        // If the middle element is greater than the rightmost element,
        // the minimum element is in the right half of the array.
        if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else {
            // Otherwise, the minimum element is in the left half or is the middle element.
            right = mid;
        }
    }

    // At the end of the loop, left and right will be pointing to the minimum element.
    return nums[left];
}

/**
 * 3. Recursive Binary Search Approach
 *
 * @param {number[]} nums The rotated sorted array.
 * @param {number} left The left index of the search range.
 * @param {number} right The right index of the search range.
 * @returns {number} The minimum element in the array.
 */
function findMinimumRecursiveBinarySearch(nums, left, right) {
    if (!nums || nums.length === 0) {
        return undefined; // Handle empty array
    }
    if (left === right) {
        return nums[left];
    }

    const mid = Math.floor((left + right) / 2);

    if (nums[mid] > nums[right]) {
        return findMinimumRecursiveBinarySearch(nums, mid + 1, right);
    } else {
        return findMinimumRecursiveBinarySearch(nums, left, mid);
    }
}

/**
 * 4. Binary Search with Duplicates Handling
 *
 * @param {number[]} nums The rotated sorted array with possible duplicates.
 * @returns {number} The minimum element in the array.
 */
function findMinimumWithDuplicates(nums) {
    if (!nums || nums.length === 0) {
        return undefined;
    }

    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        const mid = Math.floor((left + right) / 2);

        if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else if (nums[mid] < nums[right]) {
            right = mid;
        } else {
            // Handle the case with duplicates: if nums[mid] == nums[right], we can't
            // be sure which half contains the minimum, so we shrink the search space
            // by one.
            right--;
        }
    }
    return nums[left];
}

/**
 * 5. ES6 Optimized Binary Search
 *
 * @param {number[]} nums The rotated sorted array.
 * @returns {number} The minimum element.
 */
function findMinimumES6(nums) {
    if (!nums || nums.length === 0) return undefined;

    let [left, right] = [0, nums.length - 1];

    while (left < right) {
        const mid = (left + right) >>> 1; // Use bitwise for slightly faster division
        if (nums[mid] > nums[right]) left = mid + 1;
        else right = mid;
    }
    return nums[left];
}

// Example usage and console logs
const rotatedArray1 = [3, 4, 5, 1, 2];
const rotatedArray2 = [4, 5, 6, 7, 0, 1, 2];
const rotatedArray3 = [11, 13, 15, 17, 10];
const rotatedArray4 = [2, 2, 2, 0, 2]; // Example with duplicates
const rotatedArray5 = [10, 1, 10, 10, 10]; // Example with duplicates

console.log("Test Case 1:");
console.log("Input:", rotatedArray1);
console.log("Linear Search:", findMinimumLinearSearch(rotatedArray1)); // Output: 1
console.log("Binary Search:", findMinimumBinarySearch(rotatedArray1)); // Output: 1
console.log("Recursive Binary Search:", findMinimumRecursiveBinarySearch(rotatedArray1, 0, rotatedArray1.length - 1)); // Output: 1
console.log("Binary Search with Duplicates:", findMinimumWithDuplicates(rotatedArray1)); // Output: 1
console.log("ES6 Optimized Binary Search:", findMinimumES6(rotatedArray1)); // Output: 1

console.log("\nTest Case 2:");
console.log("Input:", rotatedArray2);
console.log("Linear Search:", findMinimumLinearSearch(rotatedArray2)); // Output: 0
console.log("Binary Search:", findMinimumBinarySearch(rotatedArray2)); // Output: 0
console.log("Recursive Binary Search:", findMinimumRecursiveBinarySearch(rotatedArray2, 0, rotatedArray2.length - 1)); // Output: 0
console.log("Binary Search with Duplicates:", findMinimumWithDuplicates(rotatedArray2)); // Output: 0
console.log("ES6 Optimized Binary Search:", findMinimumES6(rotatedArray2)); // Output: 0

console.log("\nTest Case 3:");
console.log("Input:", rotatedArray3);
console.log("Linear Search:", findMinimumLinearSearch(rotatedArray3)); // Output: 10
console.log("Binary Search:", findMinimumBinarySearch(rotatedArray3)); // Output: 10
console.log("Recursive Binary Search:", findMinimumRecursiveBinarySearch(rotatedArray3, 0, rotatedArray3.length - 1)); // Output: 10
console.log("Binary Search with Duplicates:", findMinimumWithDuplicates(rotatedArray3)); // Output: 10
console.log("ES6 Optimized Binary Search:", findMinimumES6(rotatedArray3)); // Output: 10

console.log("\nTest Case 4 (Duplicates):");
console.log("Input:", rotatedArray4);
console.log("Linear Search:", findMinimumLinearSearch(rotatedArray4)); // Output: 0
console.log("Binary Search:", findMinimumBinarySearch(rotatedArray4)); // Output: 0
console.log("Recursive Binary Search:", findMinimumRecursiveBinarySearch(rotatedArray4, 0, rotatedArray4.length - 1)); // Output: 0
console.log("Binary Search with Duplicates:", findMinimumWithDuplicates(rotatedArray4)); // Output: 0
console.log("ES6 Optimized Binary Search:", findMinimumES6(rotatedArray4)); // Output: 0

console.log("\nTest Case 5 (Duplicates):");
console.log("Input:", rotatedArray5);
console.log("Linear Search:", findMinimumLinearSearch(rotatedArray5));
console.log("Binary Search:", findMinimumBinarySearch(rotatedArray5));
console.log("Recursive Binary Search:", findMinimumRecursiveBinarySearch(rotatedArray5, 0, rotatedArray5.length - 1));
console.log("Binary Search with Duplicates:", findMinimumWithDuplicates(rotatedArray5));
console.log("ES6 Optimized Binary Search:", findMinimumES6(rotatedArray5));

File: C++/5-BinarySearch/Problems/3-FindMinimum-in-RotatedSortedArray/FindMinimum-in-RotatedSortedArrayPro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits> // Required for numeric_limits

using namespace std;

// Function to find the minimum element in a rotated sorted array.

// Approach 1: Linear Search
// Description:  This is the most basic approach. Iterate through the entire array
//               and keep track of the smallest element encountered.
//               While simple, it's not the most efficient, especially for large arrays.
// Time Complexity: O(n) - where n is the number of elements in the array.
// Space Complexity: O(1) - constant extra space.
int findMin_LinearSearch(const vector<int>& nums) {
    if (nums.empty()) {
        return -1; // Handle the edge case of an empty array.  Returning -1 to indicate an error.  Could also throw an exception.
    }

    int minElement = nums[0]; // Initialize with the first element.
    for (size_t i = 1; i < nums.size(); ++i) {
        if (nums[i] < minElement) {
            minElement = nums[i]; // Update minElement if a smaller element is found.
        }
    }
    return minElement;
}

// Approach 2: Using STL's min_element
// Description:  This approach leverages the C++ Standard Template Library (STL)
//               function `min_element`.  It provides a concise way to find the minimum
//               element in a range.
// Time Complexity: O(n) -  `min_element` performs a linear search.
// Space Complexity: O(1) - constant extra space.
int findMin_STL(const vector<int>& nums) {
    if (nums.empty()) {
        return -1; // Handle empty array
    }
    return *min_element(nums.begin(), nums.end());
}

// Approach 3: Binary Search (Optimized)
// Description:  This is the most efficient approach.  It utilizes the property
//               of the rotated sorted array: it's sorted in two parts.  We can
//               use binary search to find the point where the rotation occurs,
//               which is where the minimum element resides.
// Time Complexity: O(log n) -  Binary search reduces the search space by half at each step.
// Space Complexity: O(1) -  Constant extra space.
int findMin_BinarySearch(const vector<int>& nums) {
    if (nums.empty()) {
        return -1; // Handle empty array
    }

    int left = 0;
    int right = nums.size() - 1;

    // If the array is not rotated, the first element is the minimum.
    if (nums[left] <= nums[right]) {
        return nums[left];
    }

    while (left < right) {
        int mid = left + (right - left) / 2; // Prevent potential overflow.

        if (nums[mid] > nums[right]) {
            // The minimum element is in the right half.
            left = mid + 1;
        } else {
            // The minimum element is in the left half (including mid).
            right = mid;
        }
    }
    // 'left' and 'right' converge to the minimum element.
    return nums[left];
}

// Approach 4: Recursive Binary Search
// Description: A recursive implementation of the binary search approach.
//              This approach provides an alternative way to implement
//              the divide-and-conquer strategy of binary search.
// Time Complexity: O(log n)
// Space Complexity: O(log n) - Due to the recursive call stack.
int findMin_RecursiveBinarySearchHelper(const vector<int>& nums, int left, int right) {
    if (left >= right) {
        return nums[left];
    }

    int mid = left + (right - left) / 2;

     if (nums[left] <= nums[right]) { // added this condition
        return nums[left];
    }

    if (nums[mid] > nums[right]) {
        return findMin_RecursiveBinarySearchHelper(nums, mid + 1, right);
    } else {
        return findMin_RecursiveBinarySearchHelper(nums, left, mid);
    }
}

int findMin_RecursiveBinarySearch(const vector<int>& nums) {
    if (nums.empty()) {
        return -1;
    }
    return findMin_RecursiveBinarySearchHelper(nums, 0, nums.size() - 1);
}

// Approach 5:  Binary Search with Duplicate Handling
// Description: This approach handles the case where the rotated sorted array
//              contains duplicate elements.  Duplicates can make it harder to
//              determine which half of the array contains the minimum element.
// Time Complexity: O(log n) in the average case, but O(n) in the worst case
//                  (when there are many duplicate elements).
// Space Complexity: O(1)
int findMin_BinarySearchDuplicates(const vector<int>& nums) {
    if (nums.empty()) {
        return -1;
    }

    int left = 0;
    int right = nums.size() - 1;

    while (left < right) {
        int mid = left + (right - left) / 2;

        if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else if (nums[mid] < nums[right]) {
            right = mid;
        } else {
            // Handle the case where nums[mid] == nums[right].
            // We can't be sure which side the minimum is on, so we
            // reduce the search space by one.
            right--;
        }
    }
    return nums[left];
}

int main() {
    // Example usage of the different approaches.
    vector<int> rotatedArray1 = {4, 5, 6, 7, 0, 1, 2};
    vector<int> rotatedArray2 = {3, 4, 5, 1, 2};
    vector<int> rotatedArray3 = {10, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    vector<int> rotatedArray4 = {2, 2, 2, 0, 1}; // Example with duplicates
    vector<int> rotatedArray5 = {1, 1, 0, 1, 1, 1};
    vector<int> emptyArray = {};

    cout << "Rotated Array 1: ";
    for (int num : rotatedArray1) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(rotatedArray1) << endl;
    cout << "STL's min_element: " << findMin_STL(rotatedArray1) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(rotatedArray1) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(rotatedArray1) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(rotatedArray1) << endl;

    cout << "\nRotated Array 2: ";
     for (int num : rotatedArray2) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(rotatedArray2) << endl;
    cout << "STL's min_element: " << findMin_STL(rotatedArray2) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(rotatedArray2) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(rotatedArray2) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(rotatedArray2) << endl;

    cout << "\nRotated Array 3: ";
    for (int num : rotatedArray3) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(rotatedArray3) << endl;
    cout << "STL's min_element: " << findMin_STL(rotatedArray3) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(rotatedArray3) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(rotatedArray3) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(rotatedArray3) << endl;

    cout << "\nRotated Array 4 (with duplicates): ";
    for (int num : rotatedArray4) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(rotatedArray4) << endl;
    cout << "STL's min_element: " << findMin_STL(rotatedArray4) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(rotatedArray4) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(rotatedArray4) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(rotatedArray4) << endl;

    cout << "\nRotated Array 5 (with duplicates): ";
    for (int num : rotatedArray5) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(rotatedArray5) << endl;
    cout << "STL's min_element: " << findMin_STL(rotatedArray5) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(rotatedArray5) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(rotatedArray5) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(rotatedArray5) << endl;

    cout << "\nEmpty Array: ";
    for (int num : emptyArray) {
        cout << num << " ";
    }
    cout << endl;
    cout << "Linear Search: " << findMin_LinearSearch(emptyArray) << endl;
    cout << "STL's min_element: " << findMin_STL(emptyArray) << endl;
    cout << "Binary Search: " << findMin_BinarySearch(emptyArray) << endl;
    cout << "Recursive Binary Search: " << findMin_RecursiveBinarySearch(emptyArray) << endl;
    cout << "Binary Search with Duplicates: " << findMin_BinarySearchDuplicates(emptyArray) << endl;

    return 0;
}
